File: python/0518.py
# Link: https://leetcode.com/problems/coin-change-ii/description/
# Runtime: 269 ms
# Memory: 16.41 MB

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        number_of_combinations = [[0] * (amount + 1)] * (len(coins) + 1)
        for i in range(len(coins)):
            number_of_combinations[i][0] = 1
        
        for i in range(len(coins) - 1, -1, -1):
            for j in range(1, amount + 1):
                if coins[i] > j:
                    number_of_combinations[i][j] = number_of_combinations[i + 1][j]
                else:
                    number_of_combinations[i][j] = number_of_combinations[i + 1][j] + number_of_combinations[i][j - coins[i]]
        
        return number_of_combinations[i][amount]

File: c++/0518.cpp
// Link: https://leetcode.com/problems/coin-change-ii/description/
// Runtime: 39 ms
// Memory: 18.73 MB

class Solution {
public:
  int change(int amount, const std::vector<int> &coins) {
    std::vector<std::vector<int>> numberOfCombinations(
        coins.size() + 1, std::vector<int>(amount + 1));
    for (int i = 0; i < static_cast<int>(coins.size()); i++)
      numberOfCombinations[i][0] = 1;

    for (int i = static_cast<int>(coins.size()) - 1; i >= 0; i--)
      for (int j = 1; j <= amount; j++) {
        if (coins[i] > j)
          numberOfCombinations[i][j] = numberOfCombinations[i + 1][j];
        else
          numberOfCombinations[i][j] = numberOfCombinations[i + 1][j] +
                                       numberOfCombinations[i][j - coins[i]];
      }

    return numberOfCombinations[0][amount];
  }
};
